checkstyle.xml

```diff
@@ -0,0 +1,122 @@
+    </module>
+
+    <!-- No Trailing Whitespace, except on lines that only have an asterisk (e.g. Javadoc comments) -->
+    <module name="RegexpSingleline">
+        <property name="format" value="(?&lt;!\*)\s+$|\*\s\s+$"/>
+        <property name="message" value="Line has trailing whitespace"/>
+        <property name="fileExtensions" value="java, properties, cfg, css, js, xml"/>
+    </module>
+
+    <!-- Allow individual lines of code to be excluded from these rules, if they are annotated
+         with @SuppressWarnings. See also SuppressWarningsHolder below -->
+    <module name="SuppressWarningsFilter" />
+
+    <!-- Check individual Java source files for specific rules -->
+    <module name="TreeWalker">
+        <!-- Maximum line length is 120 characters -->
+        <module name="LineLength">
+            <property name="max" value="120"/>
+        </module>
+
+        <!-- Check for and report back any TODO or FIXME comments as info messages -->
+        <module name="TodoComment">
+            <property name="severity" value="info"/>
+            <property name="format" value="(TODO)|(FIXME)"/>
+        </module>
+
+        <!-- Do not report errors on any lines annotated with @SuppressWarnings -->
+        <module name="SuppressWarningsHolder"/>
+
+        <!-- ##### Import statement requirements ##### -->
+        <!-- Star imports (e.g. import java.util.*) are NOT ALLOWED -->
+        <module name="AvoidStarImport"/>
+        <!-- Redundant import statements are NOT ALLOWED -->
+        <module name="RedundantImport"/>
+        <!-- Unused import statements are NOT ALLOWED -->
+        <module name="UnusedImports"/>
+
+        <!-- ##### Javadocs requirements ##### -->
+        <!-- Requirements for Javadocs for classes/interfaces -->
+        <module name="JavadocType">
+            <!-- All public classes/interfaces MUST HAVE Javadocs -->
+            <property name="scope" value="public"/>
+            <!-- Add an exception for anonymous inner classes -->
+            <property name="excludeScope" value="anoninner"/>
+            <!-- Ignore errors related to unknown tags -->
+            <property name="allowUnknownTags" value="true"/>
+            <!-- Allow params tags to be optional -->
+            <property name="allowMissingParamTags" value="false"/>
+        </module>
+        <!-- Requirements for Javadocs for methods -->
+        <module name="JavadocMethod">
+            <!-- All public methods MUST HAVE Javadocs -->
+            <property name="scope" value="public"/>
+            <!-- Allow RuntimeExceptions to be undeclared -->
+            <property name="allowUndeclaredRTE" value="true"/>
+            <!-- Allow params, throws and return tags to be optional -->
+            <property name="allowMissingParamTags" value="true"/>
+            <property name="allowMissingThrowsTags" value="true"/>
+            <property name="allowMissingReturnTag" value="true"/>
+        </module>
```

Comment:
Regarding type, method, variable and parameter names, I think it's also a good idea to automatically enforce a naming convention. For example: type names have to start with a capital, variables should start with a lowercase letter...

```
<module name="TypeName">
    <message key="name.invalidPattern"
     value="Type name ''{0}'' must match pattern ''{1}''."/>
</module>
<module name="MemberName">
    <property name="format" value="^[a-z][a-z0-9][a-zA-Z0-9]*$"/>
    <message key="name.invalidPattern"
     value="Member name ''{0}'' must match pattern ''{1}''."/>
</module>
<module name="ParameterName">
    <property name="format" value="^[a-z]([a-z0-9][a-zA-Z0-9]*)?$"/>
    <message key="name.invalidPattern"
     value="Parameter name ''{0}'' must match pattern ''{1}''."/>
</module>
<module name="CatchParameterName">
    <property name="format" value="^[a-z]([a-z0-9][a-zA-Z0-9]*)?$"/>
    <message key="name.invalidPattern"
     value="Catch parameter name ''{0}'' must match pattern ''{1}''."/>
</module>
<module name="LocalVariableName">
    <property name="tokens" value="VARIABLE_DEF"/>
    <property name="format" value="^[a-z]([a-z0-9][a-zA-Z0-9]*)?$"/>
    <message key="name.invalidPattern"
     value="Local variable name ''{0}'' must match pattern ''{1}''."/>
</module>
<module name="ClassTypeParameterName">
    <property name="format" value="(^[A-Z][0-9]?)$|([A-Z][a-zA-Z0-9]*[T]$)"/>
    <message key="name.invalidPattern"
     value="Class type name ''{0}'' must match pattern ''{1}''."/>
</module>
<module name="MethodTypeParameterName">
    <property name="format" value="(^[A-Z][0-9]?)$|([A-Z][a-zA-Z0-9]*[T]$)"/>
    <message key="name.invalidPattern"
     value="Method type name ''{0}'' must match pattern ''{1}''."/>
</module>
<module name="MethodName">
    <property name="format" value="^[a-z][a-z0-9][a-zA-Z0-9_]*$"/>
    <message key="name.invalidPattern"
     value="Method name ''{0}'' must match pattern ''{1}''."/>
</module>
<module name="InterfaceTypeParameterName">
    <property name="format" value="(^[A-Z][0-9]?)$|([A-Z][a-zA-Z0-9]*[T]$)"/>
    <message key="name.invalidPattern"
     value="Interface type name ''{0}'' must match pattern ''{1}''."/>
</module>
```

(taken from https://github.com/checkstyle/checkstyle/blob/master/src/main/resources/google_checks.xml#L111)

Reply:
I notice that the underscore character is only present in one of these rules.  Does that restriction make sense?

I have not given much thought to the naming of type parameters in the past.  Does the requirement to end with a T reflect best practices?

I could be helpful to update this proposal with a brief explanation for the different patterns.

Reply:
Hi @terrywbrady, I've updated the regexes a bit and added comments trying to explain them.

Reply:
I have not included this suggestion at this time, as I worry it will slow down the initial effort by requiring a larger amount of code refactoring. The current concentration is on standardizing our indentations and braces to ease PR reviews (especially for DSpace 7).  However, we should consider these extra rules for the future.

Reply:
As discussed during one of the developer meetings, we agreed to not apply these rules now. Therefore I'm removing this comment.